### PR TITLE
Cache interpreter at stack

### PR DESCRIPTION
**Pull request by He-Pin**
Cache it at the stack for performance

### CODE REVIEW

**Comment by hepin1989** [comment]
`sbt.ForkMain$ForkError: org.jboss.netty.channel.ChannelException: Failed to bind to: localhost/127.0.0.1:33445
	at org.jboss.netty.bootstrap.ServerBootstrap.bind(ServerBootstrap.java:272)
	at akka.remote.transport.netty.NettyTransport.$anonfun$listen$1(NettyTransport.scala:498)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:92)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:92)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:42)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:47)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: sbt.ForkMain$ForkError: java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at org.jboss.netty.channel.socket.nio.NioServerBoss$RegisterTask.run(NioServerBoss.java:193)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.processTaskQueue(AbstractNioSelector.java:391)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:315)
	at org.jboss.netty.channel.socket.nio.NioServerBoss.run(NioServerBoss.java:42)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)`

**Comment by johanandren** [comment]
Do you have any benchmark numbers that shows an improvement?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 1]
```diff
@@ -606,6 +608,7 @@ import scala.util.control.NonFatal
 
   def runBatch(actorEventLimit: Int): Int = {
     try {
+      val interpreter = this.interpreter
```
[the remark this answers is not in the document]
Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Will do

**Comment by patriknw** [comment]
> Do you have any benchmark numbers that shows an improvement?

I don't think this will show up in benchmarks, but it's good to not access that `lazy val` more than necessary.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 2]
```diff
@@ -482,6 +483,7 @@ import scala.util.control.NonFatal
       extends BoundaryEvent {
     override def execute(eventLimit: Int): Int = {
       if (!waitingForShutdown) {
+        val interpreter = shell.interpreter
```
[the remark this answers is not in the document]
Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: in another `interpreter.isSuspended`. I rename to `interp`

**Review by raboof — APPROVED** [review]

**Review by jrudolph — CHANGES_REQUESTED** [review]
Let's not do premature optimization.

**Comment by hepin1989** [comment]
@jrudolph This kind of code is very common, as a high-performance toolkit, this is not premature, especially when you deploy on a large scale.

**Comment by jrudolph** [comment]
> @jrudolph This kind of code is very common, as a high-performance toolkit, this is not premature, especially when you deploy on a large scale.

In general, it's always a trade-off between maintainability and ultimate performance. More often than not, in the long run over-optimizing prevents incremental improvements because code has become too complex. This change seems benign but things like that accumulate and make further changes harder. And also, often, it turns out that simple code is also the fastest.

In this case, removing multiple accesses to a lazy val is an optimization that the compiler should do if needed. If we want to remove lazy val accesses here, let's remove the lazy val in the first place.

**Comment by patriknw** [comment]
closing since different opinions and no benchmark numbers showing that it's needed

**Pull request closed without merging**